### PR TITLE
Docker - Support docker pull

### PR DESCRIPTION
**Pull request by Or-Gabay**
- [x] All [tests](https://ci.appveyor.com/project/jfrog-ecosystem/build-info) passed. If this feature is not already covered by the tests, I added new tests.
-----
Support for docker pull, including:

* Fat-manifest.
* Marker layers.
* Collect build info.

Related PR:
* https://github.com/jfrog/jenkins-artifactory-plugin/pull/347

### CODE REVIEW

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java [hunk 1]
```diff
@@ -118,4 +128,25 @@ public Build execute() {
...
+
+    /**
+     * Update each layer's properties with artifactProperties.
+     */
+    private void setImageLayersProps(DockerLayers layers, ArrayListMultimap<String, String> artifactProperties, ArtifactoryBuildInfoClientBuilder buildInfoClientBuilder) throws IOException {
+        String artifactsPropsStr = buildMatrixParamsString(artifactProperties, false);
+        try (ArtifactoryBuildInfoClient buildInfoClient = buildInfoClientBuilder.build()) {
+            for (DockerLayer layer : layers.getLayers()) {
+                HttpResponse httpResponse = buildInfoClient.executeUpdateFileProperty(layer.getFullPath(), artifactsPropsStr);
+                validateResponse(httpResponse);
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:140]:
```suggestion
                validateResponse(httpResponse);
                EntityUtils.consume(httpResponse.getEntity());
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java [hunk 2]
```diff
@@ -118,4 +128,25 @@ public Build execute() {
...
+            }
+        }
+    }
+
+    private void validateResponse(HttpResponse httpResponse) throws IOException {
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:145]:
```suggestion
    private void validateSetImageLayersResponse(HttpResponse httpResponse) throws IOException {
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 3]
```diff
@@ -23,49 +21,71 @@
...
             manifest = candidateManifest;
             imagePath = candidateImagePath;
+            loadLayers();
         }
     }
 
-    private String getManifestFromArtifactory(ArtifactoryDependenciesClient dependenciesClient, String manifestPath) throws IOException {
+    // Searching for image's manifest in Artifactory. For remote repositories, fat-manifest can be found instead,
+    // therefore, we must compare the fat-manifest against our image architecture to find the right digest(represent our image's manifest digest).
+    // Using the correct digest from the fat-manifest, we are able to build the path toward manifest.json in Artifactory.
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:72]: Please covert to JavaDoc and explain each variable

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 4]
```diff
@@ -23,49 +21,71 @@
...
         HttpResponse res = null;
         try {
-            res = dependenciesClient.downloadArtifact(manifestPath);
+            res = dependenciesClient.downloadArtifact(manifestPath + "/manifest.json");
+            return IOUtils.toString(res.getEntity().getContent());
+        } catch (Exception e) {
+            if (!isRemoteRepo) {
+                throw e;
+            }
+            res = dependenciesClient.downloadArtifact(manifestPath + "/list.manifest.json");
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:82]: Please add a comment explaining this logic

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 5]
```diff
@@ -207,64 +206,83 @@ public Module generateBuildInfoModule(ArtifactoryBuildInfoClientBuilder buildInf
...
-            temp = artifactProperties.get("build.timestamp");
-            if (temp != null && temp.size() > 0) {
-                buildInfoItemsProps.setProperty("build.timestamp", temp.get(0));
+        try {
+            checkAndSetManifestAndImagePathCandidates(proxyLessManifestPath, proxyLessImagePath, dependenciesClient, isRemoteRepo);
+        } catch (IOException e) {
+            logger.error("The manifest could not be fetched from Artifactory, assuming proxy-lessess configuration - " + e.getMessage());
+            // If image path includes more than 3 slashes, Artifactory doesn't store this image under 'library',
+            // thus we should not look further.
+            int totalSlash = proxyImagePath.length() - proxyImagePath.replace("/", "").length();
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:270]:
```suggestion
            int totalSlash = StringUtils.countMatches(proxyImagePath, "/");
```

File: build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/client/ArtifactoryDependenciesClient.java [hunk 6]
```diff
@@ -194,6 +199,27 @@ public boolean isArtifactoryOSS() throws IOException {
...
+    }
+
+    public void downloadMarkerLayer(String repo, String imageName, String ImageMarkerName) throws IOException {
+        String url = artifactoryUrl + "/api/docker/" + repo + "/v2/" + imageName + "/blobs/" + toNoneMarkerLayer(ImageMarkerName);
+        this.getArtifactMetadata(url);
+    }
+
+    private String toNoneMarkerLayer(String markerLayerName) {
+        return markerLayerName.substring(0, markerLayerName.length() - ".marker".length());
+    }
```
Comment (yahavi) [inline build-info-extractor/src/main/java/org/jfrog/build/extractor/clientConfiguration/client/ArtifactoryDependenciesClient.java:221]: These methods are not related to the Dependencies client. Please move them to DockerUtils.

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 7]
```diff
@@ -93,8 +98,35 @@ private void dockerPushTest() {
         }
     }
 
+    @Test
+    @SuppressWarnings("unused")
+    private void dockerPullTest() {
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:103]:
```suggestion
    public void dockerPullTest() {
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 8]
```diff
@@ -146,35 +143,40 @@ private void setDependenciesAndArtifacts(Module buildInfoModule, Properties prop
...
-                .append("\"path\":\"").append(imagePath).append("\",\"$or\":[");
-        List<String> layersQuery = new ArrayList<String>();
+                .append("\"repo\":\"").append(Repo).append("\",\"$or\":[");
+        List<String> layersQuery = new ArrayList<>();
         for (String digest : layersDigest) {
             String shaVersion = DockerUtils.getShaVersion(digest);
             String singleFileQuery;
             if (StringUtils.equalsIgnoreCase(shaVersion, "sha1")) {
-                ;
                 singleFileQuery = String.format("{\"actual_sha1\": \"%s\"}", DockerUtils.getShaValue(digest));
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:177]: AQL strings should be as short as possible
```suggestion
                singleFileQuery = String.format("{\"actual_sha1\":\"%s\"}", DockerUtils.getShaValue(digest));
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 9]
```diff
@@ -146,35 +143,40 @@ private void setDependenciesAndArtifacts(Module buildInfoModule, Properties prop
...
             } else {
-                singleFileQuery = String.format("{\"name\": \"%s\"}", DockerUtils.digestToFileName(digest));
+                singleFileQuery = String.format("{\"name\":{\"$match\" : \"%s*\"}}", DockerUtils.digestToFileName(digest));
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:179]:
```suggestion
                singleFileQuery = String.format("{\"name\":{\"$match\":\"%s*\"}}", DockerUtils.digestToFileName(digest));
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 10]
```diff
@@ -207,64 +206,83 @@ public Module generateBuildInfoModule(ArtifactoryBuildInfoClientBuilder buildInf
...
     }
 
+    private void loadLayers() throws IOException {
+        try (ArtifactoryBuildInfoClient buildInfoClient = buildInfoClientBuilder.build();
+             ArtifactoryDependenciesClient dependenciesClient = dependenciesClientBuilder.build()) {
+            this.layers = this.getLayers(dependenciesClient, buildInfoClient);
+            List<DockerLayer> markerLayers = this.layers.getLayers().stream().filter(layer -> layer.getFileName().endsWith(".marker")).collect(Collectors.toList());
+            if (markerLayers.size() > 0) {
+                for (DockerLayer markerLayer : markerLayers) {
+                    dependenciesClient.downloadMarkerLayer(targetRepo, this.imageTag.substring(imageTag.indexOf("/") + 1, imageTag.indexOf(":")), markerLayer.getDigest());
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:226]:
```suggestion
                    dependenciesClient.downloadMarkerLayer(targetRepo, StringUtils.substringBetween(imageTag, "/", ":"), markerLayer.getDigest());

```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 11]
```diff
@@ -207,64 +206,83 @@ public Module generateBuildInfoModule(ArtifactoryBuildInfoClientBuilder buildInf
...
+                }
+                this.layers = this.getLayers(dependenciesClient, buildInfoClient);
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:228]: `this` is redundant. Please use only in getters/setters/when there is no other choice.
```suggestion
                layers = getLayers(dependenciesClient, buildInfoClient);
```

**Review by yahavi — CHANGES_REQUESTED** [review]
* Please make sure the tests are running both locally and in Appveyor before the next review cycle. Are you sure the test have to have both local and remote Docker repositories?
* Please replace `awaitSuccess` by `awaitCompletion` in DockerJavaWrapper, since `awaitSuccess` is deprecated.
* We recently remove Guava from the build-info. Please do not use `Hashing.sha1().hashString(...)`
* There are some warnings in DockerUtils.java. Please fix them.
* There are tons of same class variables between DockerPull and DockerPush. Please create a base parent class `DockerCommand`, same as npm does.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java (hunk 9 above)
Comment (yahavi) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:179]: Please also explain why the wildcard is necessary

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 12]
```diff
@@ -132,11 +131,9 @@ private void setDependenciesAndArtifacts(Module buildInfoModule, Properties prop
         for (int i = 0; i < dependencyLayerNum; i++) {
             String digest = it.next();
             DockerLayer layer = layers.getByDigest(digest);
-            HttpResponse httpResponse = propertyChangeClient.executeUpdateFileProperty(layer.getFullPath(), artifactsPropsStr);
-            validateResponse(httpResponse);
-            Dependency dependency = new DependencyBuilder().id(layer.getFileName()).sha1(layer.getSha1()).addProperties(properties).build();
+            Dependency dependency = new DependencyBuilder().id(layer.getFileName()).sha1(layer.getSha1()).build();
             dependencies.add(dependency);
-            Artifact artifact = new ArtifactBuilder(layer.getFileName()).sha1(layer.getSha1()).addProperties(properties).build();
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:139]: Can't we set custom properties on pushed docker images?

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java [hunk 13]
```diff
@@ -46,6 +41,40 @@ public static String getConfigDigest(String manifest) throws IOException {
...
+        }
+        for (JsonNode manifestInfo : manifests) {
+            JsonNode manifestInfoPlatform = manifestInfo.get("platform");
+            if (manifestInfoPlatform == null) {
+                continue;
+            }
+
+            JsonNode currOs = manifestInfoPlatform.get("os");
+            JsonNode currArch = manifestInfoPlatform.get("architecture");
+            if (os == null || arch == null) {
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java:67]: You are right, I mixed it up checking the wrong variables

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java (hunk 4 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:82]: I rewrote the overall comment.

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java (hunk 12 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:139]: Yes, but here we deal with the buildinfo.json artifacts & dependencies.

Comment (yahavi) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:139]: Where do you set the properties? I couldn't find

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java [hunk 14]
```diff
@@ -103,19 +92,43 @@ public Build execute() {
...
+     * Update each layer's properties with artifactProperties.
+     */
+    private void setImageLayersProps(DockerLayers layers, ArrayListMultimap<String, String> artifactProperties, ArtifactoryBuildInfoClientBuilder buildInfoClientBuilder) throws IOException {
+        String artifactsPropsStr = buildMatrixParamsString(artifactProperties, false);
+        try (ArtifactoryBuildInfoClient buildInfoClient = buildInfoClientBuilder.build()) {
+            for (DockerLayer layer : layers.getLayers()) {
+                HttpResponse httpResponse = buildInfoClient.executeUpdateFileProperty(layer.getFullPath(), artifactsPropsStr);
+                validateSetImageLayersResponse(httpResponse);
+                EntityUtils.consume(httpResponse.getEntity());
+
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/extractor/DockerPush.java:122]: Lets make sure we always consume the HTTP response:
```suggestion
                HttpResponse httpResponse = null;
                try {
                    httpResponse = buildInfoClient.executeUpdateFileProperty(layer.getFullPath(), artifactsPropsStr);
                    validateSetImageLayersResponse(httpResponse);
                } finally {
                    if (httpResponse != null) {
                        EntityUtils.consume(httpResponse.getEntity());
                    }
                }
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 15]
```diff
@@ -22,51 +20,83 @@
...
-            res = dependenciesClient.downloadArtifact(manifestPath);
-            return IOUtils.toString(res.getEntity().getContent());
+            res = dependenciesClient.downloadArtifact(manifestPath + "/manifest.json");
+            return IOUtils.toString(res.getEntity().getContent(), StandardCharsets.UTF_8);
+        } catch (Exception e) {
+            if (isLocalRepo) {
+                throw e;
+            }
+            res = dependenciesClient.downloadArtifact(manifestPath + "/list.manifest.json");
+            String digestsFromFatManifest = DockerUtils.getImageDigestFromFatManifest(IOUtils.toString(res.getEntity().getContent()), this.os, this.architecture);
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:93]: 1. You already have a dedicated method in DockerUtils to extract the string from the entity.
2. Unless you have local variables named os and architecture, there is no need to use the `this` keyword in methods that are no getters and setters.
```suggestion
            String digestsFromFatManifest = DockerUtils.getImageDigestFromFatManifest(entityToString(res.getEntity()), os, architecture);
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 16]
```diff
@@ -22,51 +20,83 @@
...
+            loadLayers();
         }
     }
 
-    private String getManifestFromArtifactory(ArtifactoryDependenciesClient dependenciesClient, String manifestPath) throws IOException {
+    /**
+     * Search for 'manifest.json' at 'manifestPath' in artifactory and return its content.
+     * If the target repository is not local type , fat-manifest can be found instead, which is a list of 'manifest.json' digest for one or more platforms.
+     * In order to find the right digest, we read and iterate on each digest and search for the one with the same os and arch.
+     * Using the correct digest from the fat-manifest, we are able to build the path toward manifest.json in Artifactory.
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:75]:
```suggestion
     * Search for 'manifest.json' at 'manifestPath' in artifactory and return its content.
     * If the target repository is not local type, fat-manifest can be found instead, which is a list of 'manifest.json' digest for one or more platforms.
     * In order to find the right digest, we read and iterate on each digest and search for the one with the same os and arch.
     * By using the correct digest from the fat-manifest, we are able to build the path toward manifest.json in Artifactory.
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java [hunk 17]
```diff
@@ -22,51 +20,83 @@
...
+            if (isLocalRepo) {
+                throw e;
+            }
+            res = dependenciesClient.downloadArtifact(manifestPath + "/list.manifest.json");
+            String digestsFromFatManifest = DockerUtils.getImageDigestFromFatManifest(IOUtils.toString(res.getEntity().getContent()), this.os, this.architecture);
+            if (digestsFromFatManifest.isEmpty()) {
+                throw e;
+            }
+            // Remove the tag from the pattern, and place the manifest digest instead.
+            res = dependenciesClient.downloadArtifact(manifestPath.substring(0, manifestPath.lastIndexOf("/")) + "/" + digestsFromFatManifest.replace(":", "__") + "/manifest.json");
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:98]:
```suggestion
            res = dependenciesClient.downloadArtifact(StringUtils.substringBeforeLast(manifestPath, "/") + "/" + digestsFromFatManifest.replace(":", "__") + "/manifest.json");

```

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 18]
```diff
@@ -93,8 +108,64 @@ private void dockerPushTest() {
...
+        } catch (Exception e) {
+            fail(ExceptionUtils.getStackTrace(e));
+        }
+    }
+
+    private void validatePulledBuild(Build build) {
+        assertEquals(build.getModules().size(), 1);
+        Module module = build.getModules().get(0);
+
+        assertEquals(null, module.getArtifacts());
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:155]:
```suggestion
        assertNull(module.getArtifacts());
```

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 19]
```diff
@@ -93,8 +108,64 @@ private void dockerPushTest() {
...
+                throw new IOException("The " + DOCKER_VIRTUAL_REPO + " environment variable is not set, failing docker tests.");
+            }
+            DockerPull dockerPull = new DockerPull(buildInfoClientBuilder, dependenciesClientBuilder, pullImageFromVirtual, host, dockerVirtualRepo, getUsername(), getPassword(), getLog(), Collections.emptyMap());
+            validatePulledBuild(dockerPull.execute());
+        } catch (Exception e) {
+            fail(ExceptionUtils.getStackTrace(e));
+        }
+    }
+
+    private void validatePulledBuild(Build build) {
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:151]:
```suggestion
    private void validatePulledDockerImage(Build build) {
```

File: appveyor.yml [hunk 20]
```diff
@@ -2,16 +2,7 @@ stack: jdk 8, node 8, go 1.13, python 3.7.7
...
-  BITESTS_ARTIFACTORY_URL:
-    secure: RIh0gGFDn2JAnLUEEqLsQm0ihToJ6/4LiR411QpvqDgYYE7Y4Eif2e+wYndvHXnW
-  BITESTS_ARTIFACTORY_USERNAME:
-    secure: YEB5Wiv9a2vNpUhy+MfL+A==
-  BITESTS_ARTIFACTORY_PASSWORD:
-    secure: Um8o75MQIieSavIemF4ySA==
-  BITESTS_ARTIFACTORY_DOCKER_DOMAIN:
-    secure: AoaKTKtH3eL36SmENsm51eRFVLU/BGIK+CrBvmsCqO35j3xM9++4E01YWs5CsfXy
-  BITESTS_ARTIFACTORY_DOCKER_REPO:
-    secure: KL72XB/1EBaModiGGpJF9Ge+wdbLmrDwT1lmy28YRiM=
```
Comment (yahavi) [inline appveyor.yml:14]: Please don't forget to re-add them before merging

**Review by yahavi — CHANGES_REQUESTED** [review]

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java (hunk 12 above)
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/types/DockerImage.java:139]: https://github.com/jfrog/build-info/pull/397/files#diff-f2b37fdb29ea988c82ea3a71485db026f328e2802ae871abe4ca41b4ea79f816R100

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java [hunk 21]
```diff
@@ -237,4 +281,18 @@ public static void initTempDir(File path) {
         // Set the Java temp dir system property. As a result, java will create it for us.
         System.setProperty("java.io.tmpdir", Paths.get(pathDir, "DockerJavaTemp").toString());
     }
+
+    public static String toNoneMarkerLayer(String markerLayerName) {
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java:285]: Removed this method.

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 22]
```diff
@@ -59,48 +77,156 @@ private void setUp() {
...
+
+    private String getImageId(String image) {
+        return DockerJavaWrapper.InspectImage(image, host, Collections.emptyMap(), getLog()).getId().replace(":", "__");
+    }
+
+    private String validateDomainSuffix(String domain) {
+        if (!StringUtils.endsWith(domain, "/")) {
+            domain += "/";
+        }
+        return domain;
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:225]:
```suggestion
        return StringUtils.appendIfMissing(domain, "/");
```

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 23]
```diff
@@ -59,48 +77,156 @@ private void setUp() {
...
+
+    private void validatePulledDockerImage(Build build, String image) {
+        assertEquals(build.getModules().size(), 1);
+        Module module = build.getModules().get(0);
+        assertEquals(module.getType(), "docker");
+        List<Dependency> dependencies = module.getDependencies();
+        validateImageDepedencies(dependencies, image);
+    }
+
+    private void validateImageDepedencies(List<Dependency> deps, String image) {
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:205]:
```suggestion
    private void validateImageDependencies(List<Dependency> deps, String image) {
```

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 24]
```diff
@@ -59,48 +77,156 @@ private void setUp() {
...
+        assertTrue(deps.stream().anyMatch(dep -> dep.getId().equals(imageDigest)));
+    }
+
+    private void validateImageArtifacts(List<Artifact> arts, String image) {
+        String imageDigest = getImageId(image);
+        assertTrue(arts.stream().anyMatch(art -> art.getName().equals(imageDigest)));
+    }
+
+    private String getImageId(String image) {
+        return DockerJavaWrapper.InspectImage(image, host, Collections.emptyMap(), getLog()).getId().replace(":", "__");
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:218]:
```suggestion
        String id = DockerJavaWrapper.InspectImage(image, host, Collections.emptyMap(), getLog()).getId();
        assertNotNull(id);
        return id.replace(":", "__");
```

File: build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java [hunk 25]
```diff
@@ -59,48 +77,156 @@ private void setUp() {
...
+        assertEquals(build.getModules().size(), 1);
+        Module module = build.getModules().get(0);
+        assertEquals(module.getType(), "docker");
+        List<Dependency> dependencies = module.getDependencies();
+        validateImageDepedencies(dependencies, image);
+    }
+
+    private void validateImageDepedencies(List<Dependency> deps, String image) {
+        // Latest tag may change the number of dependencies in the future.
+        assertTrue(deps.size() > 0);
```
Comment (yahavi) [inline build-info-extractor-docker/src/test/java/org/jfrog/build/extractor/docker/extractor/DockerExtractorTest.java:207]:
```suggestion
        assertFalse(deps.isEmpty());
```

File: build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java [hunk 26]
```diff
@@ -47,7 +48,44 @@ public static String getConfigDigest(String manifest) throws IOException {
...
+     * @param manifest - fat-manifest.
+     * @param os       -      image os to search.
+     * @param arch     -    arch to search.
+     * @return digest related to os and arch. If not found return an empty string.
+     * @throws IOException fat-manifest has missing 'manifest' key.
+     */
+    public static String getImageDigestFromFatManifest(String manifest, String os, String arch) throws IOException {
+        if (Strings.isNullOrEmpty(os) || Strings.isNullOrEmpty(arch)) {
+            return "";
+        }
```
Comment (yahavi) [inline build-info-extractor-docker/src/main/java/org/jfrog/build/extractor/docker/DockerUtils.java:62]:
```suggestion
        if (StringUtils.isAnyBlank(os, arch)) {
            return StringUtils.EMPTY;
        }
```

**Review by yahavi — CHANGES_REQUESTED** [review]
